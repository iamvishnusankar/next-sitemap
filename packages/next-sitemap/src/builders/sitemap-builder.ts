import type {
  IAlternateRef,
  IGoogleNewsEntry,
  IImageEntry,
  ISitemapField,
  IVideoEntry,
} from '../interface.js'
import { entityEscapedUrl } from '../utils/url.js'

/**
 * Builder class to generate xml and robots.txt
 * Returns only string values
 */
export class SitemapBuilder {
  /**
   * Create XML Template
   * @param content
   * @returns
   */
  withXMLTemplate(content: string): string {
    return `<?xml version="1.0" encoding="UTF-8"?>\n<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">\n${content}</urlset>`
  }

  /**
   * Generates sitemap-index.xml
   * @param allSitemaps
   * @returns
   */
  buildSitemapIndexXml(allSitemaps: string[]) {
    return [
      '<?xml version="1.0" encoding="UTF-8"?>',
      '<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">',
      ...(allSitemaps?.map((x) => `<sitemap><loc>${x}</loc></sitemap>`) ?? []),
      '</sitemapindex>',
    ].join('\n')
  }

  /**
   * Normalize sitemap field keys to stay consistent with <xsd:sequence> order
   * @link https://www.w3schools.com/xml/el_sequence.asp
   * @link https://github.com/iamvishnusankar/next-sitemap/issues/345
   * @param x
   * @returns
   */
  normalizeSitemapField(x: ISitemapField) {
    const { loc, lastmod, changefreq, priority, ...restProps } = x

    // Return keys in following order
    return {
      loc,
      lastmod,
      changefreq,
      priority,
      ...restProps,
    }
  }

  /**
   * Composes YYYY-MM-DDThh:mm:ssTZD date format (with TZ offset)
   * (ref: https://stackoverflow.com/a/49332027)
   * @param date
   * @private
   */
  private formatDate(date: Date | string): string {
    const d = typeof date === 'string' ? new Date(date) : date
    const z = (n) => ('0' + n).slice(-2)
    const zz = (n) => ('00' + n).slice(-3)
    let off = d.getTimezoneOffset()
    const sign = off > 0 ? '-' : '+'
    off = Math.abs(off)

    return (
      d.getFullYear() +
      '-' +
      z(d.getMonth() + 1) +
      '-' +
      z(d.getDate()) +
      'T' +
      z(d.getHours()) +
      ':' +
      z(d.getMinutes()) +
      ':' +
      z(d.getSeconds()) +
      '.' +
      zz(d.getMilliseconds()) +
      sign +
      z((off / 60) | 0) +
      ':' +
      z(off % 60)
    )
  }

  private formatBoolean(value: boolean): string {
    return value ? 'yes' : 'no'
  }

  private escapeHtml(s: string) {
    return s.replace(/[^\dA-Za-z ]/g, (c) => '&#' + c.charCodeAt(0) + ';')
  }

  /**
   * Generates sitemap.xml
   * @param fields
   * @returns
   */
  buildSitemapXml(fields: ISitemapField[]): string {
    const content = fields
      .map((x: ISitemapField) => {
        // Normalize sitemap field keys to stay consistent with <xsd:sequence> order
        const field = this.normalizeSitemapField(x)

        // Field array to keep track of properties
        const fieldArr: Array<string> = []

        // Iterate all object keys and key value pair to field-set
        for (const key of Object.keys(field)) {
          // Skip reserved keys
          if (['trailingSlash'].includes(key)) {
            continue
          }

          if (field[key]) {
            if (key === 'alternateRefs') {
              const altRefField = this.buildAlternateRefsXml(
                field.alternateRefs
              )

              fieldArr.push(altRefField)
            } else if (key === 'news') {
              if (field.news) {
                const newsField = this.buildNewsXml(field.news)
                fieldArr.push(newsField)
              }
            } else if (key === 'images') {
              if (field.images) {
                for (const image of field.images) {
                  const imageField = this.buildImageXml(image)
                  fieldArr.push(imageField)
                }
              }
            } else if (key === 'videos') {
              if (field.videos) {
                for (const video of field.videos) {
                  const videoField = this.buildVideoXml(video)
                  fieldArr.push(videoField)
                }
              }
            } else {
              fieldArr.push(`<${key}>${field[key]}</${key}>`)
            }
          }
        }

        // Append previous value and return
        return `<url>${fieldArr.join('')}</url>\n`
      })
      .join('')

    return this.withXMLTemplate(content)
  }

  /**
   * Generate alternate refs.xml
   * @param alternateRefs
   * @returns
   */
  buildAlternateRefsXml(alternateRefs: Array<IAlternateRef> = []): string {
    return alternateRefs
      .map((alternateRef) => {
        return `<xhtml:link rel="alternate" hreflang="${alternateRef.hreflang}" href="${alternateRef.href}"/>`
      })
      .join('')
  }

  /**
   * Generate Google News sitemap entry
   * @param news
   * @returns string
   */
  buildNewsXml(news: IGoogleNewsEntry): string {
    // using array just because it looks more structured
    return [
      `<news:news>`,
      ...[
        `<news:publication>`,
        ...[
          `<news:name>${this.escapeHtml(news.publicationName)}</news:name>`,
          `<news:language>${news.publicationLanguage}</news:language>`,
        ],
        `</news:publication>`,
        `<news:publication_date>${this.formatDate(
          news.date
        )}</news:publication_date>`,
        `<news:title>${this.escapeHtml(news.title)}</news:title>`,
      ],
      `</news:news>`,
    ]
      .filter(Boolean)
      .join('')
  }

  /**
   * Generate Image sitemap entry
   * @param image
   * @returns string
   */
  buildImageXml(image: IImageEntry): string {
    // using array just because it looks more structured
    return [
      `<image:image>`,
      ...[
        `<image:loc>${entityEscapedUrl(image.loc.href)}</image:loc>`,
        image.caption &&
          `<image:caption>${this.escapeHtml(image.caption)}</image:caption>`,
        image.title &&
          `<image:title>${this.escapeHtml(image.title)}</image:title>`,
        image.geoLocation &&
          `<image:geo_location>${this.escapeHtml(
            image.geoLocation
          )}</image:geo_location>`,
        image.license &&
          `<image:license>${entityEscapedUrl(
            image.license.href
          )}</image:license>`,
      ],
      `</image:image>`,
    ]
      .filter(Boolean)
      .join('')
  }

  /**
   * Generate Video sitemap entry
   * @param video
   * @returns string
   */
  buildVideoXml(video: IVideoEntry): string {
    // using array just because it looks more structured
    return [
      `<video:video>`,
      ...[
        `<video:title>${this.escapeHtml(video.title)}</video:title>`,
        `<video:thumbnail_loc>${entityEscapedUrl(
          video.thumbnailLoc.href
        )}</video:thumbnail_loc>`,
        `<video:description>${this.escapeHtml(
          video.description
        )}</video:description>`,
        video.contentLoc &&
          `<video:content_loc>${entityEscapedUrl(
            video.contentLoc.href
          )}</video:content_loc>`,
        video.playerLoc &&
          `<video:player_loc>${entityEscapedUrl(
            video.playerLoc.href
          )}</video:player_loc>`,
        video.duration && `<video:duration>${video.duration}</video:duration>`,
        video.viewCount &&
          `<video:view_count>${video.viewCount}</video:view_count>`,
        video.tag && `<video:tag>${this.escapeHtml(video.tag)}</video:tag>`,
        video.rating &&
          `<video:rating>${video.rating
            .toFixed(1)
            .replace(',', '.')}</video:rating>`,
        video.expirationDate &&
          `<video:expiration_date>${this.formatDate(
            video.expirationDate
          )}</video:expiration_date>`,
        video.publicationDate &&
          `<video:publication_date>${this.formatDate(
            video.publicationDate
          )}</video:publication_date>`,
        typeof video.familyFriendly !== 'undefined' &&
          `<video:family_friendly>${this.formatBoolean(
            video.familyFriendly
          )}</video:family_friendly>`,
        typeof video.requiresSubscription !== 'undefined' &&
          `<video:requires_subscription>${this.formatBoolean(
            video.requiresSubscription
          )}</video:requires_subscription>`,
        typeof video.live !== 'undefined' &&
          `<video:live>${this.formatBoolean(video.live)}</video:live>`,
        video.restriction &&
          `<video:restriction relationship="${video.restriction.relationship}">${video.restriction.content}</video:restriction>`,
        video.platform &&
          `<video:platform relationship="${video.platform.relationship}">${video.platform.content}</video:platform>`,
        video.uploader &&
          `<video:uploader${
            video.uploader.info &&
            ` info="${entityEscapedUrl(video.uploader.info.href)}"`
          }>${this.escapeHtml(video.uploader.name)}</video:uploader>`,
      ],
      `</video:video>`,
    ]
      .filter(Boolean)
      .join('')
  }
}
