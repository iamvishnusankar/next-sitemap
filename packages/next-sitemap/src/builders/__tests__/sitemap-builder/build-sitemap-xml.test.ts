import { SitemapBuilder } from '../../sitemap-builder.js'
import { XMLValidator } from 'fast-xml-parser'

describe('SitemapBuilder', () => {
  test('snapshot test to exclude undefined values from final sitemap', () => {
    // Builder instance
    const builder = new SitemapBuilder()

    // Build content
    const content = builder.buildSitemapXml([
      {
        loc: 'https://example.com',
        lastmod: undefined,
      },
      {
        loc: 'https://example.com',
        lastmod: 'some-value',
        alternateRefs: [
          {
            href: 'https://example.com/en',
            hreflang: 'en',
          },
          {
            href: 'https://example.com/fr',
            hreflang: 'fr',
          },
        ],
      },
    ])
    const isValid = XMLValidator.validate(content)
    expect(isValid).toBe(true)
    // Expect the generated sitemap to match snapshot.
    expect(content).toMatchSnapshot()
  })
  test('snapshot test for google news sitemap', () => {
    // Builder instance
    const builder = new SitemapBuilder()

    // Build content
    const content = builder.buildSitemapXml([
      {
        loc: 'https://example.com',
        news: {
          title: 'Companies A, B in Merger Talks',
          date: new Date('2008-01-02T00:00:00.000+01:00'),
          publicationLanguage: 'en',
          publicationName: 'The Example Times',
        },
      },
    ])
    const isValid = XMLValidator.validate(content)
    expect(isValid).toBe(true)
    // Expect the generated sitemap to match snapshot.
    expect(content).toMatchSnapshot()
  })
  test('snapshot test for image sitemap', () => {
    // Builder instance
    const builder = new SitemapBuilder()

    // Build content
    const content = builder.buildSitemapXml([
      {
        loc: 'https://example.com',
        images: [
          {
            loc: new URL('https://example.com'),
          },
          {
            caption: 'Image caption & description',
            geoLocation: 'Prague, Czech Republic',
            license: new URL('https://example.com'),
            loc: new URL('https://example.com'),
            title: 'Image title',
          },
        ],
      },
    ])
    const isValid = XMLValidator.validate(content)
    expect(isValid).toBe(true)
    // Expect the generated sitemap to match snapshot.
    expect(content).toMatchSnapshot()
  })

  test('image sitemap generates correctly encoded results', () => {
    // Builder instance
    const builder = new SitemapBuilder()

    // Build content
    const content = builder.buildSitemapXml([
      {
        loc: 'https://example.com',
        images: [
          {
            loc: new URL('https://example.com?ref=test&test=1'),
          },
        ],
      },
    ])

    // Expect the generated sitemap to match snapshot.
    const isValid = XMLValidator.validate(content)
    expect(isValid).toBe(true)
    expect(content).toMatchSnapshot()
  })
  test('snapshot test for video sitemap', () => {
    // Builder instance
    const builder = new SitemapBuilder()

    // Build content
    const content = builder.buildSitemapXml([
      {
        loc: 'https://example.com',
        videos: [
          {
            title: 'Video title',
            contentLoc: new URL('https://example.com'),
            description: 'Video description',
            thumbnailLoc: new URL('https://example.com'),
          },
          {
            title: 'Grilling steaks for summer',
            contentLoc: new URL('https://example.com'),
            description:
              'Alkis shows you how to get perfectly done steaks every time',
            thumbnailLoc: new URL('https://example.com'),
            duration: 600,
            expirationDate: new Date('2030-03-02T00:00:00.000+01:00'),
            familyFriendly: true,
            live: false,
            platform: {
              relationship: 'allow',
              content: 'web',
            },
            playerLoc: new URL('https://example.com'),
            publicationDate: new Date('2020-04-20T00:00:00.000+02:00'),
            rating: 1,
            requiresSubscription: false,
            restriction: {
              relationship: 'deny',
              content: 'CZ',
            },
            tag: 'video',
            uploader: {
              name: 'John Doe',
              info: new URL('https://example.com'),
            },
            viewCount: 1234,
          },
        ],
      },
    ])
    const isValid = XMLValidator.validate(content)
    expect(isValid).toBe(true)
    // Expect the generated sitemap to match snapshot.
    expect(content).toMatchSnapshot()
  })
})
