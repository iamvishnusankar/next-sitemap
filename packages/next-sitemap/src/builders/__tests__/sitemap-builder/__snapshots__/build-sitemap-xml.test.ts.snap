// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`SitemapBuilder snapshot test to exclude undefined values from final sitemap 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
<url><loc>https://example.com</loc></url>
<url><loc>https://example.com</loc><lastmod>some-value</lastmod><xhtml:link rel="alternate" hreflang="en" href="https://example.com/en"/><xhtml:link rel="alternate" hreflang="fr" href="https://example.com/fr"/></url>
</urlset>"
`;

exports[`SitemapBuilder snapshot test for google news sitemap 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
<url><loc>https://example.com</loc><news:news><news:publication><news:name>The Example Times</news:name><news:language>en</news:language></news:publication><news:publication_date>2008-01-01T23:00:00.000+00:00</news:publication_date><news:title>Companies A&#44; B in Merger Talks</news:title></news:news></url>
</urlset>"
`;

exports[`SitemapBuilder snapshot test for image sitemap 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
<url><loc>https://example.com</loc><image:image><image:loc>https://example.com/</image:loc></image:image><image:image><image:loc>https://example.com/</image:loc><image:caption>Image caption &#38; description</image:caption><image:title>Image title</image:title><image:geo_location>Prague&#44; Czech Republic</image:geo_location><image:license>https://example.com/</image:license></image:image></url>
</urlset>"
`;

exports[`SitemapBuilder image sitemap generates correctly encoded results 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
<url><loc>https://example.com</loc><image:image><image:loc>https://example.com/?ref=test&amp;test=1</image:loc></image:image></url>
</urlset>"
`;

exports[`SitemapBuilder snapshot test for video sitemap 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:news="http://www.google.com/schemas/sitemap-news/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml" xmlns:mobile="http://www.google.com/schemas/sitemap-mobile/1.0" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
<url><loc>https://example.com</loc><video:video><video:title>Video title</video:title><video:thumbnail_loc>https://example.com/</video:thumbnail_loc><video:description>Video description</video:description><video:content_loc>https://example.com/</video:content_loc></video:video><video:video><video:title>Grilling steaks for summer</video:title><video:thumbnail_loc>https://example.com/</video:thumbnail_loc><video:description>Alkis shows you how to get perfectly done steaks every time</video:description><video:content_loc>https://example.com/</video:content_loc><video:player_loc>https://example.com/</video:player_loc><video:duration>600</video:duration><video:view_count>1234</video:view_count><video:tag>video</video:tag><video:rating>1.0</video:rating><video:expiration_date>2030-03-01T23:00:00.000+00:00</video:expiration_date><video:publication_date>2020-04-19T22:00:00.000+00:00</video:publication_date><video:family_friendly>yes</video:family_friendly><video:requires_subscription>no</video:requires_subscription><video:live>no</video:live><video:restriction relationship="deny">CZ</video:restriction><video:platform relationship="allow">web</video:platform><video:uploader info="https://example.com/">John Doe</video:uploader></video:video></url>
</urlset>"
`;
